File: sdk/java/src/main/java/io/readylayer/services/PolicyService.java
package io.readylayer.services;

import io.readylayer.ReadyLayerClient;
import io.readylayer.models.*;
import lombok.RequiredArgsConstructor;

import java.net.http.HttpRequest;
import java.util.concurrent.CompletableFuture;

/**
 * Service for policy operations.
 */
@RequiredArgsConstructor
public class PolicyService {
    
    private final ReadyLayerClient client;
    
    /**
     * List all policy packs.
     */
    public PolicyPackListResponse list(ListOptions options) {
        StringBuilder path = new StringBuilder("/policies?");
        addQueryParams(path, options);
        
        HttpRequest request = client.requestBuilder(path.toString())
                .GET()
                .build();
        
        return client.execute(request, PolicyPackListResponse.class);
    }
    
    /**
     * List policy packs with default pagination.
     */
    public PolicyPackListResponse list() {
        return list(ListOptions.builder()
                .pagination(PaginationParams.builder().build())
                .build());
    }
    
    /**
     * Get a specific policy pack.
     */
    public PolicyPack get(String packId) {
        HttpRequest request = client.requestBuilder("/policies/" + packId)
                .GET()
                .build();
        
        return client.execute(request, PolicyPack.class);
    }
    
    /**
     * Create a new policy pack.
     */
    public PolicyPack create(CreatePolicyPackRequest request) {
        HttpRequest httpRequest = client.requestBuilder("/policies")
                .POST(HttpRequest.BodyPublishers.ofString(client.toJson(request)))
                .build();
        
        return client.execute(httpRequest, PolicyPack.class);
    }
    
    /**
     * Update a policy pack.
     */
    public PolicyPack update(String packId, UpdatePolicyPackRequest request) {
        HttpRequest httpRequest = client.requestBuilder("/policies/" + packId)
                .PUT(HttpRequest.BodyPublishers.ofString(client.toJson(request)))
                .build();
        
        return client.execute(httpRequest, PolicyPack.class);
    }
    
    /**
     * Delete a policy pack.
     */
    public void delete(String packId) {
        HttpRequest request = client.requestBuilder("/policies/" + packId)
                .DELETE()
                .build();
        
        client.execute(request, Void.class);
    }
    
    /**
     * List all rules in a policy pack.
     */
    public PolicyRuleListResponse listRules(String packId) {
        HttpRequest request = client.requestBuilder("/policies/" + packId + "/rules")
                .GET()
                .build();
        
        return client.execute(request, PolicyRuleListResponse.class);
    }
    
    /**
     * Add a rule to a policy pack.
     */
    public PolicyRule createRule(String packId, CreatePolicyRuleRequest request) {
        HttpRequest httpRequest = client.requestBuilder("/policies/" + packId + "/rules")
                .POST(HttpRequest.BodyPublishers.ofString(client.toJson(request)))
                .build();
        
        return client.execute(httpRequest, PolicyRule.class);
    }
    
    /**
     * Update a rule in a policy pack.
     */
    public PolicyRule updateRule(String packId, String ruleId, UpdatePolicyRuleRequest request) {
        HttpRequest httpRequest = client.requestBuilder("/policies/" + packId + "/rules/" + ruleId)
                .PUT(HttpRequest.BodyPublishers.ofString(client.toJson(request)))
                .build();
        
        return client.execute(httpRequest, PolicyRule.class);
    }
    
    /**
     * Delete a rule from a policy pack.
     */
    public void deleteRule(String packId, String ruleId) {
        HttpRequest request = client.requestBuilder("/policies/" + packId + "/rules/" + ruleId)
                .DELETE()
                .build();
        
        client.execute(request, Void.class);
    }
    
    /**
     * Validate policy syntax.
     */
    public PolicyValidationResult validate(ValidatePolicyRequest request) {
        HttpRequest httpRequest = client.requestBuilder("/policies/validate")
                .POST(HttpRequest.BodyPublishers.ofString(client.toJson(request)))
                .build();
        
        return client.execute(httpRequest, PolicyValidationResult.class);
    }
    
    /**
     * List available policy templates.
     */
    public PolicyTemplateListResponse listTemplates() {
        HttpRequest request = client.requestBuilder("/policies/templates")
                .GET()
                .build();
        
        return client.execute(request, PolicyTemplateListResponse.class);
    }
    
    private void addQueryParams(StringBuilder path, ListOptions options) {
        if (options == null) return;
        
        boolean first = true;
        
        if (options.getOrganizationId() != null) {
            path.append(first ? "" : "&").append("organizationId=").append(options.getOrganizationId());
            first = false;
        }
        if (options.getRepositoryId() != null) {
            path.append(first ? "" : "&").append("repositoryId=").append(options.getRepositoryId());
            first = false;
        }
        
        PaginationParams pagination = options.getEffectivePagination();
        path.append(first ? "" : "&")
                .append("limit=").append(pagination.getEffectiveLimit())
                .append("&offset=").append(pagination.getEffectiveOffset());
    }
}

File: sdk/csharp/src/Services/PolicyService.cs
using ReadyLayer.SDK.Models;

namespace ReadyLayer.SDK.Services;

/// <summary>
/// Service for policy pack and rule operations.
/// </summary>
public class PolicyService
{
    private readonly ReadyLayerClient _client;

    internal PolicyService(ReadyLayerClient client)
    {
        _client = client;
    }

    /// <summary>
    /// Lists all policy packs.
    /// </summary>
    /// <param name="params">Optional filtering and pagination parameters.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>List of policy packs with pagination information.</returns>
    public async Task<PolicyPackListResponse> ListPacksAsync(
        ListPolicyPacksParams? @params = null,
        CancellationToken cancellationToken = default)
    {
        @params ??= new ListPolicyPacksParams();
        return await _client.GetAsync<PolicyPackListResponse>(
            "policies",
            @params.ToQueryParams(),
            cancellationToken);
    }

    /// <summary>
    /// Gets a specific policy pack by ID.
    /// </summary>
    /// <param name="packId">The policy pack ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The policy pack details.</returns>
    public async Task<PolicyPack> GetPackAsync(
        string packId,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(packId))
            throw new ArgumentException("Policy pack ID is required", nameof(packId));

        return await _client.GetAsync<PolicyPack>(
            $"policies/{Uri.EscapeDataString(packId)}",
            cancellationToken: cancellationToken);
    }

    /// <summary>
    /// Creates a new policy pack.
    /// </summary>
    /// <param name="request">The policy pack creation request.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The created policy pack.</returns>
    public async Task<PolicyPack> CreatePackAsync(
        CreatePolicyPackRequest request,
        CancellationToken cancellationToken = default)
    {
        if (request == null)
            throw new ArgumentNullException(nameof(request));

        return await _client.PostAsync<CreatePolicyPackRequest, PolicyPack>(
            "policies",
            request,
            cancellationToken);
    }

    /// <summary>
    /// Updates a policy pack.
    /// </summary>
    /// <param name="packId">The policy pack ID.</param>
    /// <param name="request">The policy pack update request.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The updated policy pack.</returns>
    public async Task<PolicyPack> UpdatePackAsync(
        string packId,
        UpdatePolicyPackRequest request,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(packId))
            throw new ArgumentException("Policy pack ID is required", nameof(packId));
        if (request == null)
            throw new ArgumentNullException(nameof(request));

        return await _client.PutAsync<UpdatePolicyPackRequest, PolicyPack>(
            $"policies/{Uri.EscapeDataString(packId)}",
            request,
            cancellationToken);
    }

    /// <summary>
    /// Deletes a policy pack.
    /// </summary>
    /// <param name="packId">The policy pack ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task DeletePackAsync(
        string packId,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(packId))
            throw new ArgumentException("Policy pack ID is required", nameof(packId));

        await _client.DeleteAsync(
            $"policies/{Uri.EscapeDataString(packId)}",
            cancellationToken);
    }

    /// <summary>
    /// Lists all rules in a policy pack.
    /// </summary>
    /// <param name="packId">The policy pack ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>List of policy rules.</returns>
    public async Task<PolicyRuleListResponse> ListRulesAsync(
        string packId,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(packId))
            throw new ArgumentException("Policy pack ID is required", nameof(packId));

        return await _client.GetAsync<PolicyRuleListResponse>(
            $"policies/{Uri.EscapeDataString(packId)}/rules",
            cancellationToken: cancellationToken);
    }

    /// <summary>
    /// Adds a rule to a policy pack.
    /// </summary>
    /// <param name="packId">The policy pack ID.</param>
    /// <param name="request">The rule creation request.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The created policy rule.</returns>
    public async Task<PolicyRule> CreateRuleAsync(
        string packId,
        CreatePolicyRuleRequest request,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(packId))
            throw new ArgumentException("Policy pack ID is required", nameof(packId));
        if (request == null)
            throw new ArgumentNullException(nameof(request));

        return await _client.PostAsync<CreatePolicyRuleRequest, PolicyRule>(
            $"policies/{Uri.EscapeDataString(packId)}/rules",
            request,
            cancellationToken);
    }

    /// <summary>
    /// Updates a rule in a policy pack.
    /// </summary>
    /// <param name="packId">The policy pack ID.</param>
    /// <param name="ruleId">The rule ID.</param>
    /// <param name="request">The rule update request.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The updated policy rule.</returns>
    public async Task<PolicyRule> UpdateRuleAsync(
        string packId,
        string ruleId,
        UpdatePolicyRuleRequest request,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(packId))
            throw new ArgumentException("Policy pack ID is required", nameof(packId));
        if (string.IsNullOrWhiteSpace(ruleId))
            throw new ArgumentException("Rule ID is required", nameof(ruleId));
        if (request == null)
            throw new ArgumentNullException(nameof(request));

        return await _client.PutAsync<UpdatePolicyRuleRequest, PolicyRule>(
            $"policies/{Uri.EscapeDataString(packId)}/rules/{Uri.EscapeDataString(ruleId)}",
            request,
            cancellationToken);
    }

    /// <summary>
    /// Deletes a rule from a policy pack.
    /// </summary>
    /// <param name="packId">The policy pack ID.</param>
    /// <param name="ruleId">The rule ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task DeleteRuleAsync(
        string packId,
        string ruleId,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(packId))
            throw new ArgumentException("Policy pack ID is required", nameof(packId));
        if (string.IsNullOrWhiteSpace(ruleId))
            throw new ArgumentException("Rule ID is required", nameof(ruleId));

        await _client.DeleteAsync(
            $"policies/{Uri.EscapeDataString(packId)}/rules/{Uri.EscapeDataString(ruleId)}",
            cancellationToken);
    }

    /// <summary>
    /// Validates policy YAML/JSON syntax and configuration.
    /// </summary>
    /// <param name="source">The policy source to validate.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The validation result.</returns>
    public async Task<PolicyValidationResult> ValidateAsync(
        string source,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(source))
            throw new ArgumentException("Source is required", nameof(source));

        return await _client.PostAsync<ValidatePolicyRequest, PolicyValidationResult>(
            "policies/validate",
            new ValidatePolicyRequest { Source = source },
            cancellationToken);
    }

    /// <summary>
    /// Lists available policy templates.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>List of policy templates.</returns>
    public async Task<PolicyTemplateListResponse> ListTemplatesAsync(
        CancellationToken cancellationToken = default)
    {
        return await _client.GetAsync<PolicyTemplateListResponse>(
            "policies/templates",
            cancellationToken: cancellationToken);
    }
}
